--- a/include/linux/netdevice.h	2021-06-16 14:55:42.724976114 -0600
+++ b/include/linux/netdevice.h	2021-06-16 15:15:42.038719467 -0600
@@ -186,6 +186,9 @@
 	unsigned long	tx_window_errors;
 	unsigned long	rx_compressed;
 	unsigned long	tx_compressed;
+
+	/* Unknown extra variable Ubiquiti introduced in 1.10.0-12 */
+	unsigned long 	extravar;
 };
 
 
@@ -1796,6 +1799,9 @@
 	atomic_long_t		tx_dropped;
 	atomic_long_t		rx_nohandler;
 
+	/* Unknown extra variable Ubiquiti introduced in 1.10.0-12 */
+	atomic_long_t 		extravar;
+
 	/* Stats to monitor link on/off, flapping */
 	atomic_t		carrier_up_count;
 	atomic_t		carrier_down_count;
--- a/include/uapi/linux/if_link.h	2021-06-16 14:55:42.848304733 -0600
+++ b/include/uapi/linux/if_link.h	2021-06-16 15:13:57.121955175 -0600
@@ -38,6 +38,9 @@
 	__u32	tx_compressed;
 
 	__u32	rx_nohandler;		/* dropped, no handler found	*/
+
+	/* Unknown extra variable Ubiquiti introduced in 1.10.0-12 */
+	__u32   extravar;
 };
 
 /* The main device statistics structure */
@@ -73,6 +76,9 @@
 	__u64	tx_compressed;
 
 	__u64	rx_nohandler;		/* dropped, no handler found	*/
+
+	/* Unknown extra variable Ubiquiti introduced in 1.10.0-12 */
+	__u64   extravar;
 };
 
 /* The struct should be in sync with struct ifmap */
