--- a/scripts/gen_initramfs_list.sh
+++ b/scripts/gen_initramfs_list.sh
@@ -55,14 +55,26 @@ default_initramfs() {
 		nod /dev/console 0600 0 0 c 5 1
 		dir /root 0700 0 0
 		# file /kinit usr/kinit/kinit 0755 0 0
 		# slink /init kinit 0755 0 0
 	EOF
 }
 
+list_openwrt_initramfs() {
+	:
+}
+
+openwrt_initramfs() {
+	# make sure that /dev/console exists
+	cat <<-EOF >> ${output}
+		dir /dev 0755 0 0
+		nod /dev/console 0600 0 0 c 5 1
+	EOF
+}
+
 filetype() {
 	local argv1="$1"
 
 	# symlink test must come before file test
 	if [ -L "${argv1}" ]; then
 		echo "slink"
 	elif [ -f "${argv1}" ]; then
@@ -173,14 +185,16 @@ dir_filelist() {
 	srcdir=$(echo "$1" | sed -e 's://*:/:g')
 	dirlist=$(find "${srcdir}" -printf "%p %m %U %G\n")
 
 	# If $dirlist is only one line, then the directory is empty
 	if [  "$(echo "${dirlist}" | wc -l)" -gt 1 ]; then
 		${dep_list}print_mtime "$1"
 
+		${dep_list}openwrt_initramfs
+
 		echo "${dirlist}" | \
 		while read x; do
 			${dep_list}parse ${x}
 		done
 	fi
 }
 
